gh-41431: Add `datetime.time.strptime()` and `datetime.date.strptime()` (#120752)
python/cpython: Doc/library/datetime.rst

   .. versionadded:: 3.8

.. classmethod:: date.strptime(date_string, format)

   Return a :class:`.date` corresponding to *date_string*, parsed according to
   *format*. This is equivalent to::

     date(*(time.strptime(date_string, format)[0:3]))

   :exc:`ValueError` is raised if the date_string and format
   can't be parsed by :func:`time.strptime` or if it returns a value which isn't a
   time tuple.  See also :ref:`strftime-strptime-behavior` and
   :meth:`date.fromisoformat`.

>    .. note

Author: Is it fine to have this note twice in the documentation?

Reviewer: I think documentation experts will know better than me, but I think there's a way to define a substitution for this, like this:

```
|format-note|

.. |format-note| replace::
    If *format* specifies a day of month without a year a :exc:`DeprecationWarning` is emitted. This is to avoid a quadrenial leap year bug in code seeking to parse only a month and day, as the default year used when unspecified in the format is not a leap year. Such *format* values may raise an error in Python 3.15. The workaround is to always include a year in your *format*. If parsing *date_string* values that do not have a year, explicitly add a year that is a leap year before parsing:
    .. doctest::
        
        >> from datetime import date
        >> date_string = "02/29"
        >> when = date.strptime(f"{date_string};1984", "%m/%d;%Y")  # Avoids leap year bug.
        >> when.strftime("%B %d") # doctest: +SKIP
        'February 29'

.. versionadded:: 3.14
```

Then you put `|format-note|` in the other place as well.

Two other options:
  1. move this to a small fragment file (seems like overkill), and use an `..include` directive with it.
  2. put the note in one place, then link to it in the other place (e.g. "See the note on :func:`datetime.strptime`").

I don't think there's a simple way to do it with a parameter, though.

Author: I was just wondering if it was a problem it was mentioned twice. Do I link to `datetime.strptime` or is this fine?

      Previously, this method only supported formats that could be emitted by
      :meth:`time.isoformat()`.

.. classmethod:: time.strptime(date_string, format)

   Return a :class:`.time` corresponding to *date_string*, parsed according to
   *format*.

   If *format* does not contain microseconds or timezone information, this is equivalent to::

>      time(*(time.strptime(date_string, format)[3:6]))

Reviewer 2: I'd use a simpler example. This is one is too convoluted IMO.

Author: It's adapted from https://docs.python.org/3.14/library/datetime.html#datetime.datetime.strptime

>    :exc:`ValueError` is raised if the date_string and format

Reviewer 2: ```suggestion
   :exc:`ValueError` is raised if the *date_string* and *format*
```

Reviewer 2: I know that it is to keep consistent with `datetime.datetime` but I think it's better to make it that way to make it consistent with the rest of the docs rather (as you can see on the current docs, those two names are emphasized in the next paragraphs so I think it's something that was omitted rather than desired).

Author: This is a fine change in my opinion. I mostly just copied the code and documentation from `datetime.strptime()`.

>    can't be parsed by :func:`time.strptime` or if it returns a value which isn't a

Reviewer 2: ```suggestion
   cannot be parsed by :func:`time.strptime` or if it returns a value which is not a
```

>    time tuple.  See also :ref:`strftime-strptime-behavior` and

Reviewer 2: Maybe use `.. seealso::` directive?

Author: Let's keep it consistent with the rest.

``strftime(format)`` method, to create a string representing the time under the
control of an explicit format string.

Conversely, the :meth:`date.strptime`, :meth:`datetime.strptime` and

> :meth:`time.strptime`,  class methods create an object from a string

Reviewer 2: ```suggestion
:meth:`time.strptime` class methods create an object from a string
```
